### ISSUE TITLE
STACK question:You seem to be missing * characters

### ISSUE THREAD

**Issue by miroi** [maths/moodle-qtype_stack#652]
Hello,

in my STACK question I defined variables as c0SO2,crSO2 etc , but got this strange message upon expression evaluation:

"Your last answer was interpreted as follows: (c0SO2-crSO2)^2/((c0O2-0.5*(c0SO2-crSO2))*crSO2^2)

This answer is invalid. You seem to be missing * characters. Perhaps you meant to type (c0*S*O2-crSO2)^2/((c0*O2-0.5*(c0*S*O2-crSO2))*crSO2^2)."

Is there a way to deactivate this uncorrect suggestion "...Perhaps you meant to type (c0*S*O2" ? I  I prefer to combine letters and numbers in variable names, like c0_SO2, or c_r_SO2 ...

Comments please ?

**Comment by sangwinc** [comment]
Miroi,
I think it would be easiest if you email me your question (C.J.Sangwin@ed.ac.uk), and I'll take a look at the whole thing.
Chris Sangwin

**Comment by miroi** [comment]
Hello CHris,

here is the zipped xml:
[test-Chemická informatika-(LFUPJŠ2016-628) ZATIAĽ TESTOVACIA OTAZKA-20201109-1229.zip](https://github.com/maths/moodle-qtype_stack/files/5510089/test-Chemicka.informatika-.LFUPJS2016-628.ZATIAL.TESTOVACIA.OTAZKA-20201109-1229.zip)

a here is the print-screen:
![stack-printscreen](https://user-images.githubusercontent.com/1047336/98536026-62e47200-2287-11eb-8284-8e51c37c5610.png)

**Comment by aharjula** [comment]
Chris,

Yet another case where an option to drop [this one](https://github.com/maths/moodle-qtype_stack/blob/c252205afc925eba2d6d45dc11cba02b2e7044f4/stack/input/inputbase.class.php#L838) out seems to be required. Though why it breaks between `S` and `O` as well might also require additional options, and the corrective parser would also need to not do [this](https://github.com/maths/moodle-qtype_stack/blob/master/stack/maximaparser/corrective_parser.php#L154). Again one of those places where the initial assumptions of the shapes of variables and common errors do not quite fit when changing subjects and will lead to an [explosion of options](https://github.com/maths/moodle-qtype_stateful/blob/master/stateful/input2/inputs/algebraic.input.php#L155-L157), well at least some work is being done to get this done but as stated this is far in the future and will not help here. Though the filters of 4.3 do make this simpler to solve, even now, with an extra option if one so wishes...

**Comment by miroi** [comment]
Dear experts,

FYI: we, chemistry teachers, like well named variables,  like c0_SO2 , meaning initial concentration of SO2 (sulphur dioxide), or cr_SO2, meaning equilibrium concentration of SO2.

**Comment by sangwinc** [comment]
Just to clarify, you trying to type in a chemical formula?!  STACK is designed for mathematics, and this adopts conventions etc. which are separate from chemistry. Over the years, this has probably been the most popular request!   I've always tried to solve the problem of assessment by being completely honest to the subject.   

Rather than "patch up" the * system, I think a better approach would be to create a correct grammar for chemistry formulae, and to encode the meaning of what a student types.  This will, of course, come with proper feedback to students.  E.g. you could reject impossible combinations on the basis of chemistry.  We did work on this with scientific units.  

I'd be happy to discuss with chemists how this might work, but it is a lot more than patching *s, of course.  But if I start to patch I'm confident the chemistry colleagues will not be satisfied until we have the "job done".............. 

I think a "chemistry" input type is needed really.

Chris

**Comment by miroi** [comment]
Dear Chris,
no, no chemical formulas - just ordinary variables for chemical calculus (or physics calculus). It is about allowing more flexibility in naming variables from various fields (chemistry, physics, chemical/mechanical engineering..). I like variables with names like **c0_SO2**, but I can also use "acceptable" names instead, like **c0**.

**Comment by sangwinc** [comment]
Ok!  That is a relief! We will add more flexibility in a future version.
Chris

**Comment by miroi** [comment]
FYI: how to replace chemical formulas - instead of formulas represented as "complex variables" use references to formulas represented as simple variables, see https://moodle.org/mod/forum/discuss.php?d=415805#p1675954 .

**Comment by miroi** [comment]
Dear Chris ! A year has passed, is there any news concerning handling complex variable names, like **c0_SO2** ?

**Comment by sangwinc** [comment]
No, I'm really sorry but we've been doing other, much more mathematical things over the last year!  If someone from chemistry wants to come forward to help then we will probably be able to offer support much quicker.
